### PR TITLE
Refactor couch_httpd for before_response hook

### PR DESCRIPTION
**Pull request by jaydoane**
The main idea is to reduce duplicate code, and to create a single `do/5` function where the `before_response` hook is invoked, in addition to common logging, etc. I didn't know how to create funs for mochiweb parameterized module "methods", so I wrote simple wrappers ending in "_" to avoid collisions with existing functions.

@iilyak, can you LMK what you think of this effort so far? If you think it is an improvement, I could do a similar refactor of chttpd.erl

### CODE REVIEW

File: src/couch_httpd.erl [hunk 1]
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
...
-    Headers2 = couch_httpd_cors:cors_headers(Req, Headers1),
-    Resp = MochiReq:start_response({Code, Headers2}),
-    case MochiReq:get(method) of
+start_response(Req, Code, Headers) ->
+    start_response_length(Req, Code, Headers, undefined).
+
+start_response_length(Req, Code, Headers0, Length) ->
+    couch_log:notice("**** start_response_length", []),
+    Headers = start_response_headers(Req, Headers0, Length),
+    {Resp,_,_,_} = do(fun start_response_/2, Req, Code, Headers, undefined),
```
Comment (iilyak) [inline src/couch_httpd.erl]: do is not descriptive enough to convey the purpose.

File: src/couch_httpd.erl [hunk 2]
```diff
@@ -761,13 +776,14 @@ send_json(Req, Code, Headers, Value) ->
     AllHeaders = add_headers(Req, maybe_add_default_headers(Req, Headers)),
     send_response(Req, Code, AllHeaders, {json, Value}).
 
-add_headers(#httpd{mochi_req=MochiReq}=Req, Headers) ->
-    Headers1 = http_1_0_keep_alive(MochiReq, Headers),
-
-    Headers2 = Headers1 ++ server_header() ++
-               couch_httpd_auth:cookie_auth_header(Req, Headers1),
+add_headers(Req, Headers0) ->
```
Comment (iilyak) [inline src/couch_httpd.erl:769]: You are changing the order of headers here. This is alright. But please make sure that all tests pass. I remember seeing few tests which do not use `?assertEqual(lists:usort(Expected), lists:usort(Headers))`. But rely on literal equivalence (`?assertEqual(Expected, Headers)`).

File: src/couch_httpd.erl [hunk 3]
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
...
+    couch_log:notice("**** start_response_length", []),
+    Headers = start_response_headers(Req, Headers0, Length),
+    {Resp,_,_,_} = do(fun start_response_/2, Req, Code, Headers, undefined),
+    case method(Resp) of
         'HEAD' -> throw({http_head_abort, Resp});
         _ -> ok
     end,
     {ok, Resp}.
 
+start_response_headers(Req, Headers0, Length) ->
```
Comment (iilyak) [inline src/couch_httpd.erl]: Would it be more readable:

```
start_response_headers(Req, Headers0, undefined) ->
    basic_headers(Req, Headers0);
start_response_headers(Req, Headers0, Length) ->
   Headers = basic_headers(Req, Headers0),
   include_content_length_header(Headers, Length).
```

Feel free to ignore.

File: src/couch_httpd.erl [hunk 4]
```diff
@@ -648,32 +648,32 @@ log_request(#httpd{mochi_req=MochiReq,peer=Peer}=Req, Code) ->
...
-    CookieHeader = couch_httpd_auth:cookie_auth_header(Req, Headers),
-    Headers1 = Headers ++ server_header() ++ CookieHeader,
-    Headers2 = couch_httpd_cors:cors_headers(Req, Headers1),
-    Resp = MochiReq:start_response({Code, Headers2}),
-    case MochiReq:get(method) of
+start_response(Req, Code, Headers) ->
+    start_response_length(Req, Code, Headers, undefined).
+
+start_response_length(Req, Code, Headers0, Length) ->
+    couch_log:notice("**** start_response_length", []),
```
Comment (iilyak) [inline src/couch_httpd.erl]: Don't forget to remove log statements before merge.

File: src/couch_httpd.erl (hunk 1 above)
Comment (iilyak) [inline src/couch_httpd.erl]: Would you mind adding spaces after comma in left hand side of match expression?
I.e. `{Resp, _, _, _}`.

File: src/couch_httpd.erl [hunk 5]
```diff
@@ -720,26 +738,23 @@ last_chunk(Resp) ->
...
-    log_request(Req, Code),
-    couch_stats:increment_counter([couchdb, httpd_status_codes, Code]),
-
-    Body = make_body({Type, Value}),
+send_response(Req, Code0, Headers0, {Type, Value}) ->
+    couch_log:notice("**** send_response", []),
+    Body0 = make_body({Type, Value}),
+    Headers1 = add_headers(Req, Headers0),
+    {Resp, Code, _, Body} = do(fun respond_/2, Req, Code0, Headers1, Body0),
+    maybe_log(Code, Body),
```
Comment (iilyak) [inline src/couch_httpd.erl]: How about naming it `maybe_log_response`?

**Comment by jaydoane** [comment]
@iilyak, I'm currently attempting to fix failing couchdb_cors_tests and couchdb_http_proxy_tests. There seem to be some subtle issues with this refactor which are causing a substantial number of failures, as you predicted :(

**Comment by jaydoane** [comment]
@ilyak, I made all the changes you recommended that were still relevant after bug fixes. All tests now pass. Shall I squash into one commit?

**Comment by iilyak** [comment]
@jaydoane: Please do. We might use this PR as a backup plan.

**Comment by jaydoane** [comment]
@iilyak: squashed.

**Comment by jaydoane** [comment]
Closing in favor https://github.com/apache/couchdb-couch/pull/150

**Pull request closed without merging**